### PR TITLE
Align input memory format and grad memory format for GroupNorm backward

### LINKED ISSUE #92166: Recent PyTorch nightly builds breaks DALLE2_pytorch model in Torchbench
### 🐛 Describe the bug

Example run: https://github.com/pytorch/benchmark/actions/runs/3896093654/jobs/6671547864

This is reproducible on PT 20230112 CUDA 11.6 (fbbb19599a1d162e5927542ed251fd2ba63d5163):

```
$ python run.py DALLE2_pytorch -d cuda -t train
Running train method from DALLE2_pytorch on cuda in eager mode with input batch size 4.
Traceback (most recent call last):
  File "/fsx/users/xzhao9/benchmark/run.py", line 338, in <module>
    run_one_step(test, model=m, export_metrics_file=export_metrics_file,
  File "/fsx/users/xzhao9/benchmark/run.py", line 96, in run_one_step
    func()
  File "/fsx/users/xzhao9/benchmark/torchbenchmark/util/model.py", line 277, in invoke
    self.train()
  File "/fsx/users/xzhao9/benchmark/torchbenchmark/models/DALLE2_pytorch/__init__.py", line 104, in train
    loss.backward()
  File "/data/home/xzhao9/cluster/miniconda3/envs/py310/lib/python3.10/site-packages/torch/_tensor.py", line 488, in backward
    torch.autograd.backward(
  File "/data/home/xzhao9/cluster/miniconda3/envs/py310/lib/python3.10/site-packages/torch/autograd/__init__.py", line 197, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: view size is not compatible with input tensor's size and stride (at least one dimension spans across two contiguous subspaces). Use .reshape(...) instead.
```

This problem exists in 20230101, but doesn't exist in 20221222 (https://github.com/pytorch/benchmark/actions/runs/3760224486/jobs/6390700236).

~~I suspect this is the root cause: https://github.com/pytorch/pytorch/pull/91029, since it modifies the same file as the log message code, happened between these two days, and mentioned it is BC-breaking.~~

### Versions

The root cause commit happened between 20221222 and 20230101 nightly release. But the bug exists from 20230101 onwards.

### PR DESCRIPTION
**Pull request by CaoE**
Fixes the skipped part of the test on https://github.com/pytorch/pytorch/pull/92671. Align the input memory format and the grad memory format for GroupNorm backward.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -1171,7 +1171,7 @@
...
+  input, weight, bias: "GradMode::is_enabled() || grads[1].defined() || grads[2].defined() ? infinitely_differentiable_native_group_norm_backward(grads[0], grads[1], grads[2], input, result1, result2, weight, N, C, HxW, group, eps, grad_input_mask) : (grads[0].defined() ? native_group_norm_backward_symint(grads[0].device().is_xpu() ? grads[0] : grads[0].contiguous(grads[0].device().is_cpu() ? input.suggest_memory_format() : c10::MemoryFormat::Contiguous), input.device().is_xpu() ? input : input.contiguous(input.device().is_cpu() ? input.suggest_memory_format() : c10::MemoryFormat::Contiguous), result1, result2, weight, N, C, HxW, group, grad_input_mask) : std::tuple<Tensor, Tensor, Tensor>())"
```
Comment (jgong5) [inline tools/autograd/derivatives.yaml:1174]: Curious why don't we do the `contiguous` call on `grads[0]` (`dY`) inside the kernel of each device? This way, we don't have to put such a complicated expression (very hard to read indeed) in the derivatives.yaml. Also, a change for a particular device backend won't impact another.

**Review by jgong5 — APPROVED** [review]
Fine with the change but check my comment on whether we can simplify the expression in the derivatives.yaml.

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (CaoE) [reply tools/autograd/derivatives.yaml:1177]: Yes, we should do the contiguous call inside kernels instead of derivatives.yaml.

**Comment by CaoE** [comment]
@malfet Could you please review this PR ? Thank you !

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (malfet) [reply tools/autograd/derivatives.yaml:1177]: Plus one on making this more readable in this or in followup PRs

File: aten/src/ATen/native/group_norm.cpp [hunk 2]
```diff
@@ -120,10 +120,13 @@ std::tuple<Tensor, Tensor, Tensor> native_group_norm_backward(
       at::borrow_from_optional_tensor(gamma_opt);
   const Tensor& gamma = *gamma_maybe_owned;
 
-  bool mixed_type = is_mixed_type(X, mean, rstd);
+  bool mixed_type = is_mixed_type(X, mean, rstd) || is_mixed_type(dY, mean, rstd);
```
Comment (malfet) [inline aten/src/ATen/native/group_norm.cpp:123]: Hmm, how this change is related  to the memory format problem?

File: aten/src/ATen/native/group_norm.cpp [hunk 3]
```diff
@@ -120,10 +120,13 @@ std::tuple<Tensor, Tensor, Tensor> native_group_norm_backward(
...
   if (mixed_type) {
     check_mixed_data_type(X, mean, rstd);
+    check_mixed_data_type(dY, mean, rstd);
```
Comment (malfet) [inline aten/src/ATen/native/group_norm.cpp:126]: Same as the above

File: aten/src/ATen/native/group_norm.cpp [hunk 4]
```diff
@@ -120,10 +120,13 @@ std::tuple<Tensor, Tensor, Tensor> native_group_norm_backward(
...
   }
+  auto memory_format = X.device().is_cpu() ?
+      X.suggest_memory_format() : at::MemoryFormat::Contiguous;
```
Comment (malfet) [inline aten/src/ATen/native/group_norm.cpp:129]: Hmm, why just cpu? `xpu` should be the same, shouldn't it?

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (malfet) [reply tools/autograd/derivatives.yaml:1177]: Can you please explain why it's ok to use `input.suggested_memory_format()` for grads?
And wouldn't it essentially undo all the perf work you did in the previous PR, as we would always modify grads memory format for the backward pass (is we did not allocate it with channels last)

**Review by malfet — COMMENTED** [review]
IMO it would be good to move the complex logic that currently exists in `derivatives.yaml` into the source code and add some embedded write-up on how it supposed to behave. 

I.e. I think problem with the original PR (and may be with the entire series of PRs) was that inputs were in channels last format, but layer properties were still contiguous, but treated as channels last. (I might be wrong, have not looked into the code enough)

File: aten/src/ATen/native/group_norm.cpp (hunk 2 above)
Comment (CaoE) [reply aten/src/ATen/native/group_norm.cpp:123]: It is not related to memory format, and just make sure that X and dY are in bfloat16 and parameters are in float type when it hit mixed data type.

File: aten/src/ATen/native/group_norm.cpp (hunk 3 above)
Comment (CaoE) [reply aten/src/ATen/native/group_norm.cpp:126]: Same as the above

File: aten/src/ATen/native/group_norm.cpp (hunk 4 above)
Comment (CaoE) [reply aten/src/ATen/native/group_norm.cpp:129]: This is because we just add channels last support on CPU.

**Comment by CaoE** [comment]
> Can you please explain why it's ok to use input.suggested_memory_format() for grads?
And wouldn't it essentially undo all the perf work you did in the previous PR, as we would always modify grads memory format for the backward pass (is we did not allocate it with channels last)

The PR https://github.com/pytorch/pytorch/pull/89485 add channels last support for group norm backward, which needs both grad input (grads[0]) and input (X) to be channels last. For the case 1 :  GroupNorm gets channels last input and generate channels last output in forward and GroupNorm backward gets channels last grad input and generate channels last grad output. The PR https://github.com/pytorch/pytorch/pull/89485 works well. 
For the case 2: the propagation of channels last is broken (mentioned in https://github.com/pytorch/pytorch/issues/92166) e.g. GroupNorm gets channels first input but GroupNorm backward get channels last grad input, we need align the memory format according to input first, which will not undo the perf work for the case 1.

File: aten/src/ATen/native/group_norm.cpp (hunk 2 above)
Comment (malfet) [reply aten/src/ATen/native/group_norm.cpp:123]: Can you please move those to a separate PR and I'll approve it right away?

File: aten/src/ATen/native/group_norm.cpp (hunk 4 above)
Comment (malfet) [reply aten/src/ATen/native/group_norm.cpp:129]: Please correct me if I'm wrong, but as I read the code, this change affects behavior for XPU: it used to just bypass tensor in whatever format it was allocated for CPU and XPU, but with your changes memory layout is converted to a contiguous to contiguous for XPU.

**Comment by malfet** [comment]
@CaoE, there are more than 2 cases. One should not make any assumptions about tensor layout for either forward or backward passes. I.e. sometimes the same operator can get inputs as channels first during the first call, but as channels last during the second, and it should cast them to the most appropriate format. I might be wrong, but it feels like current CPU implementation treats `mean` and `rstd` tensors layout as to be in the same, as input, so if the same op is used with channels first and channels last layout it might produce incorrect results.

**Comment by CaoE** [comment]
>  as input, so if the same op is used with channels first and channels last layout it might produce incorrect results.

Yes, if the layout of input and grad input are not the same it will produce incorrect results.

File: aten/src/ATen/native/group_norm.cpp (hunk 4 above)
Comment (CaoE) [reply aten/src/ATen/native/group_norm.cpp:129]: `native_group_norm_backward_symint(grads[0].device().is_xpu() ? grads[0] : grads[0].contiguous(), input.device().is_xpu() ? input : input.contiguous(), result1, result2, weight, N, C, HxW, group, grad_input_mask) : std::tuple<Tensor, Tensor, Tensor>())`  

`dX = at::native::empty_like(
        X,
        c10::nullopt /* dtype */,
        c10::nullopt /* layout */,
        c10::nullopt /* device */,
        c10::nullopt /* pin_memory */,
        at::MemoryFormat::Contiguous);`

The original version of GroupNorm backward shown above bypasses grads[0] and input in whatever format it was allocated for XPU, and always generate contiguous dX. 
I want to keep the  behavior for XPU in this PR and also make sure that:  input, grad[0] and dX are memory layout aligned and support both channels last and channels first on CPU.

File: aten/src/ATen/native/group_norm.cpp (hunk 2 above)
Comment (CaoE) [reply aten/src/ATen/native/group_norm.cpp]: Moved the dtype check

File: aten/src/ATen/native/group_norm.cpp (hunk 4 above)
Comment (CaoE) [reply aten/src/ATen/native/group_norm.cpp:128]: The original version, i.e., before https://github.com/pytorch/pytorch/pull/89485 merged, force the memory format to be contiguous for dX: `dX = at::native::empty_like( X, c10::nullopt /* dtype */, c10::nullopt /* layout */, c10::nullopt /* device */, c10::nullopt /* pin_memory */, at::MemoryFormat::Contiguous);`

**Review by jgong5 — APPROVED** [review]

**Comment by CaoE** [comment]
@malfet Could you please review this again ?  Thank you.

**Comment by CaoE** [comment]
@malfet Could you please review this PR ?

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**